philippemnoel/whist
# memory locking on mac!

**Ticket(s) Closed**

- Closes #6923

**Description**
This PR adds code to use patched mimalloc as a substitute for system malloc, so we can mlock our allocations.
It also includes a function to mlock statics, which is ad-hoc and requires more testing/possibly a better method. I am extremely open to suggestions.

See https://www.notion.so/whisthq/Mimalloc-and-mlock-Implementation-c2e7d80ce8dc406981928bd0e4560b89.

**Implementation**

**Documentation & Tests Added**

**Testing Instructions**

Pull whisthq/mimalloc, check out the `serina/mlock_error` branch, and build that using the building instructions. Copy the mimalloc.o file to the `build/lib/etc/mimalloc/etc folder` on the protocol side

On the protocol side:
- make sure you compile with `-D USE_MIMALLOC=ON`
- To verify that mlocking is working as intended, you can turn on `MLOCK_LOG` in the mimalloc side; you can also check that Whist's memory usage is higher than dev Whist
- To check that this is actually improving performance:
  - turn on PLOT_UDP_RECV_GAP in `debug_flags.h`
  - turn on LOG_DATA_FOR_PLOTTER in `log_statistics.h` and change other settings as you wish
  - stress your memory
  - use `plotter.py`

**Serina's ABABABAB test (suggested by Yancey)**
with mlock:
<img width="575" alt="Screen Shot 2022-09-26 at 4 11 03 PM" src="https://user-images.githubusercontent.com/13754255/192370588-b56447f4-b2ac-42d9-ba46-096bc222803e.png">
dev:
<img width="568" alt="Screen Shot 2022-09-26 at 4 11 20 PM" src="https://user-images.githubusercontent.com/13754255/192370667-ac9645ee-ccb9-4b0f-aa09-f3342bfc67a7.png">
with mlock:
<img width="595" alt="Screen Shot 2022-09-26 at 4 11 50 PM" src="https://user-images.githubusercontent.com/13754255/192370712-edf67778-e680-4368-b366-217594020af3.png">
dev:
<img width="575" alt="Screen Shot 2022-09-26 at 4 12 07 PM" src="https://user-images.githubusercontent.com/13754255/192370789-21e6b67c-7122-4d6b-bb68-de4031088d4f.png">
with mlock:
<img width="585" alt="Screen Shot 2022-09-26 at 4 12 22 PM" src="https://user-images.githubusercontent.com/13754255/192370861-fc702840-502d-4e9c-9bca-0fab37c4381a.png">
dev:
<img width="584" alt="Screen Shot 2022-09-26 at 4 12 37 PM" src="https://user-images.githubusercontent.com/13754255/192370894-5a04d6cc-f8bd-4c95-ad00-d64a82a17d1e.png">
with mlock:
<img width="582" alt="Screen Shot 2022-09-26 at 4 12 54 PM" src="https://user-images.githubusercontent.com/13754255/192370930-eb02150f-fd31-4f6d-8ea0-8fa303afd27e.png">
dev:
<img width="594" alt="Screen Shot 2022-09-26 at 4 13 29 PM" src="https://user-images.githubusercontent.com/13754255/192371023-fbd8428b-9a23-4f8f-a183-98028bfef499.png">



**PR Checklist**

- [ ] Did the PR author fully test this PR end-to-end?
- [ ] Did one PR reviewer fully test this PR end-to-end?
- [ ] Did one PR reviewer conduct a thorough code design review?
